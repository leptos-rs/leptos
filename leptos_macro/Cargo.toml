[package]
name = "leptos_macro"
version = { workspace = true }
edition = "2021"
authors = ["Greg Johnston"]
license = "MIT"
repository = "https://github.com/leptos-rs/leptos"
description = "view macro for the Leptos web framework."
readme = "../README.md"

[lib]
proc-macro = true

[dependencies]
attribute-derive = { version = "0.8", features = ["syn-full"] }
cfg-if = "1"
html-escape = "0.2"
itertools = "0.11"
prettyplease = "0.2.4"
proc-macro-error = { version = "1", default-features = false }
proc-macro2 = "1"
quote = "1"
syn = { version = "2", features = ["full"] }
rstml = "0.11.0"
leptos_hot_reload = { workspace = true }
server_fn_macro = { workspace = true }
convert_case = "0.6.0"
uuid = { version = "1", features = ["v4"] }
tracing = "0.1.37"

[dev-dependencies]
log = "0.4"
typed-builder = "0.18"
trybuild = "1"
leptos = { path = "../leptos" }
insta = "1.29"
serde = "1"

[features]
csr = []
hydrate = []
ssr = ["server_fn_macro/ssr"]
nightly = ["server_fn_macro/nightly"]
tracing = []
experimental-islands = []
trace-component-props = []

[package.metadata.cargo-all-features]
denylist = ["nightly", "tracing", "trace-component-props"]
skip_feature_sets = [["csr", "hydrate"], ["hydrate", "csr"], ["hydrate", "ssr"]]
