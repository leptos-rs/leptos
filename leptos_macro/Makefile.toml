extend = { path = "../cargo-make/main.toml" }

[tasks.test]
clear = true
dependencies = [
    "test-all",
    "test-leptos_macro-example",
    "doc-leptos_macro-example",
]

[tasks.test-leptos_macro-example]
description = "Tests the leptos_macro/example to check if macro handles doc comments correctly"
command = "cargo"
args = ["+nightly-2024-01-29", "test", "--doc"]
cwd = "example"
install_crate = false

[tasks.doc-leptos_macro-example]
description = "Docs the leptos_macro/example to check if macro handles doc comments correctly"
command = "cargo"
args = ["+nightly-2024-01-29", "doc"]
cwd = "example"
install_crate = false
