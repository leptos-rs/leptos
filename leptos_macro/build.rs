use rustc_version::{version_meta, Channel};

fn main() {
    // Set cfg flags depending on release channel
    if matches!(version_meta().unwrap().channel, Channel::Nightly) {
        println!("cargo:rustc-cfg=rustc_nightly");
    }
}
