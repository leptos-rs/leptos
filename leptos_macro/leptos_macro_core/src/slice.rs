extern crate proc_macro;

use proc_macro::TokenStream;
use quote::{quote, ToTokens};
use syn::{
    parse::{Parse, ParseStream},
    parse_macro_input,
    punctuated::Punctuated,
    Token,
};

struct SliceMacroInput {
    root: syn::Ident,
    path: Punctuated<syn::Member, Token![.]>,
}

impl Parse for SliceMacroInput {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let root: syn::Ident = input.parse()?;
        input.parse::<Token![.]>()?;
        // do not accept trailing punctuation
        let path: Punctuated<syn::Member, Token![.]> =
            Punctuated::parse_separated_nonempty(input)?;

        if path.is_empty() {
            return Err(input.error("expected identifier"));
        }

        if !input.is_empty() {
            return Err(input.error("unexpected token"));
        }

        Ok(Self { root, path })
    }
}

impl ToTokens for SliceMacroInput {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let root = &self.root;
        let path = &self.path;

        tokens.extend(quote! {
            ::leptos::reactive::computed::create_slice(
                #root,
                |st: &_| st.#path.clone(),
                |st: &mut _, n| st.#path = n
            )
        })
    }
}

pub fn slice_impl(tokens: TokenStream) -> TokenStream {
    let input = parse_macro_input!(tokens as SliceMacroInput);
    input.into_token_stream().into()
}
