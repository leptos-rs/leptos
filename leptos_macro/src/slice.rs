extern crate proc_macro;

use proc_macro::TokenStream;
use proc_macro2::Ident;
use quote::quote;
use syn::{
    parse::{Parse, ParseStream},
    parse_macro_input,
    punctuated::Punctuated,
    token::Dot,
    Token, Type,
};

struct SliceMacroInput {
    pub root: Ident,
    pub path: Punctuated<Type, Dot>,
}

impl Parse for SliceMacroInput {
    fn parse(input: ParseStream) -> Result<Self, syn::Error> {
        let root = syn::Ident::parse(input)?;
        let _dot = <Token![.]>::parse(input)?;
        let path = input.parse_terminated(Type::parse, Token![.])?;

        if path.is_empty() {
            return Err(syn::Error::new(input.span(), "Expected identifier"));
        }

        if path.trailing_punct() {
            return Err(syn::Error::new(
                input.span(),
                "Unexpected trailing `.`",
            ));
        }

        Ok(SliceMacroInput { root, path })
    }
}

impl From<SliceMacroInput> for TokenStream {
    fn from(val: SliceMacroInput) -> Self {
        let root = val.root;
        let path = val.path;

        quote! {
            ::leptos::create_slice(
                #root,
                |st: &_| st.#path.clone(),
                |st: &mut _, n| st.#path = n
            )
        }
        .into()
    }
}

pub fn slice_impl(tokens: TokenStream) -> TokenStream {
    let input = parse_macro_input!(tokens as SliceMacroInput);
    input.into()
}
