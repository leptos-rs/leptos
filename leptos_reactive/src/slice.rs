use crate::{
    create_memo, IntoSignalSetter, RwSignal, Signal, SignalSetter,
    SignalUpdate, SignalWith,
};

/// Derives a reactive slice of an [`RwSignal`](crate::RwSignal).
///
/// Slices have the same guarantees as [`Memo`s](crate::Memo):
/// they only emit their value when it has actually been changed.
///
/// Slices need a getter and a setter, and you must make sure that
/// the setter and getter only touch their respective field and nothing else.
/// They optimally should not have any side effects.
///
/// You can use slices whenever you want to react to only parts
/// of a bigger signal. The prime example would be state management,
/// where you want all state variables grouped together, but also need
/// fine-grained signals for each or some of these variables.
/// In the example below, setting an auth token will only trigger
/// the token signal, but none of the other derived signals.
/// ```
/// # use leptos_reactive::*;
/// # let runtime = create_runtime();
///
/// // some global state with independent fields
/// #[derive(Default, Clone, Debug)]
/// struct GlobalState {
///     count: u32,
///     name: String,
/// }
///
/// let state = create_rw_signal(GlobalState::default());
///
/// // `create_slice` lets us create a "lens" into the data
/// let (count, set_count) = create_slice(
///     // we take a slice *from* `state`
///     state,
///     // our getter returns a "slice" of the data
///     |state| state.count,
///     // our setter describes how to mutate that slice, given a new value
///     |state, n| state.count = n,
/// );
///
/// // this slice is completely independent of the `count` slice
/// // neither of them will cause the other to rerun
/// let (name, set_name) = create_slice(
///     // we take a slice *from* `state`
///     state,
///     // our getter returns a "slice" of the data
///     |state| state.name.clone(),
///     // our setter describes how to mutate that slice, given a new value
///     |state, n| state.name = n,
/// );
///
/// create_effect(move |_| {
///     // note: in the browser, use leptos::log! instead
///     println!("name is {}", name.get());
/// });
/// create_effect(move |_| {
///     println!("count is {}", count.get());
/// });
///
/// // setting count only causes count to log, not name
/// set_count.set(42);
///
/// // setting name only causes name to log, not count
/// set_name.set("Bob".into());
///
/// # runtime.dispose();
/// ```
#[track_caller]
pub fn create_slice<T, O, S>(
    signal: RwSignal<T>,
    getter: impl Fn(&T) -> O + Clone + Copy + 'static,
    setter: impl Fn(&mut T, S) + Clone + Copy + 'static,
) -> (Signal<O>, SignalSetter<S>)
where
    O: PartialEq,
{
    (
        create_read_slice(signal, getter),
        create_write_slice(signal, setter),
    )
}

/// Takes a memoized, read-only slice of a signal. This is equivalent to the
/// read-only half of [`create_slice`].
#[track_caller]
pub fn create_read_slice<T, O>(
    signal: RwSignal<T>,
    getter: impl Fn(&T) -> O + Clone + Copy + 'static,
) -> Signal<O>
where
    O: PartialEq,
{
    create_memo(move |_| signal.with(getter)).into()
}

/// Creates a setter to access one slice of a signal. This is equivalent to the
/// write-only half of [`create_slice`].
#[track_caller]
pub fn create_write_slice<T, O>(
    signal: RwSignal<T>,
    setter: impl Fn(&mut T, O) + Clone + Copy + 'static,
) -> SignalSetter<O> {
    let setter = move |value| signal.update(|x| setter(x, value));
    setter.into_signal_setter()
}
