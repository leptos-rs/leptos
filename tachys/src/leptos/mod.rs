use crate::{
    hydration::Cursor,
    renderer::Renderer,
    view::{
        Mountable, Position, PositionState, Render, RenderHtml, ToTemplate,
    },
};
use leptos_reactive::{create_render_effect, Effect, SignalDispose};

mod class;
mod style;

impl<F, V> ToTemplate for F
where
    F: Fn() -> V,
    V: ToTemplate,
{
    fn to_template(
        buf: &mut String,
        class: &mut String,
        style: &mut String,
        position: &mut Position,
    ) {
        // FIXME this seems wrong
        V::to_template(buf, class, style, position)
    }
}

impl<F, V, R> Render<R> for F
where
    F: Fn() -> V + 'static,
    V: Render<R>,
    V::State: 'static,
    R: Renderer,
{
    type State = Effect<V::State>;

    fn build(self) -> Self::State {
        create_render_effect(move |prev| {
            let value = self();
            if let Some(mut state) = prev {
                value.rebuild(&mut state);
                state
            } else {
                value.build()
            }
        })
    }

    #[track_caller]
    fn rebuild(self, state: &mut Self::State) {
        crate::log(&format!(
            "[REBUILDING EFFECT] Is this a mistake? {}",
            std::panic::Location::caller(),
        ));
        let old_effect = std::mem::replace(state, self.build());
        old_effect.dispose();
    }
}

impl<F, V, R> RenderHtml<R> for F
where
    F: Fn() -> V + 'static,
    V: RenderHtml<R>,
    V::State: 'static,
    R: Renderer + 'static,
    R::Node: Clone,
    R::Element: Clone,
{
    const MIN_LENGTH: usize = V::MIN_LENGTH;

    fn to_html_with_buf(self, buf: &mut String, position: &PositionState) {
        let value = self();
        value.to_html_with_buf(buf, position);
    }

    fn hydrate<const FROM_SERVER: bool>(
        self,
        cursor: &Cursor<R>,
        position: &PositionState,
    ) -> Self::State {
        let cursor = cursor.clone();
        let position = position.clone();
        create_render_effect(move |prev| {
            let value = self();
            if let Some(mut state) = prev {
                value.rebuild(&mut state);
                state
            } else {
                value.hydrate::<FROM_SERVER>(&cursor, &position)
            }
        })
    }
}

impl<M, R> Mountable<R> for Effect<M>
where
    M: Mountable<R> + 'static,
    R: Renderer,
{
    fn unmount(&mut self) {
        self.with_value_mut(|value| {
            if let Some(value) = value {
                value.unmount()
            }
        });
    }

    fn mount(
        &mut self,
        parent: &<R as Renderer>::Element,
        marker: Option<&<R as Renderer>::Node>,
    ) {
        self.with_value_mut(|value| {
            if let Some(state) = value {
                state.mount(parent, marker);
            }
        });
    }

    fn insert_before_this(
        &self,
        parent: &<R as Renderer>::Element,
        child: &mut dyn Mountable<R>,
    ) -> bool {
        self.with_value_mut(|value| {
            value
                .as_mut()
                .map(|value| value.insert_before_this(parent, child))
        })
        .flatten()
        .unwrap_or(false)
    }
}
/*
#[cfg(test)]
mod tests {
    use crate::{
        html::element::{button, main, HtmlElement},
        renderer::mock_dom::MockDom,
        view::Render,
    };
    use leptos_reactive::{create_runtime, RwSignal, SignalGet, SignalSet};

    #[test]
    fn create_dynamic_element() {
        let rt = create_runtime();
        let count = RwSignal::new(0);
        let app: HtmlElement<_, _, _, MockDom> =
            button((), move || count.get().to_string());
        let el = app.build();
        assert_eq!(el.el.to_debug_html(), "<button>0</button>");
        rt.dispose();
    }

    #[test]
    fn update_dynamic_element() {
        let rt = create_runtime();
        let count = RwSignal::new(0);
        let app: HtmlElement<_, _, _, MockDom> =
            button((), move || count.get().to_string());
        let el = app.build();
        assert_eq!(el.el.to_debug_html(), "<button>0</button>");
        count.set(1);
        assert_eq!(el.el.to_debug_html(), "<button>1</button>");
        rt.dispose();
    }

    #[test]
    fn update_dynamic_element_among_siblings() {
        let rt = create_runtime();
        let count = RwSignal::new(0);
        let app: HtmlElement<_, _, _, MockDom> = main(
            (),
            button(
                (),
                ("Hello, my ", move || count.get().to_string(), " friends."),
            ),
        );
        let el = app.build();
        assert_eq!(
            el.el.to_debug_html(),
            "<main><button>Hello, my 0 friends.</button></main>"
        );
        count.set(42);
        assert_eq!(
            el.el.to_debug_html(),
            "<main><button>Hello, my 42 friends.</button></main>"
        );
        rt.dispose();
    }
}
 */
