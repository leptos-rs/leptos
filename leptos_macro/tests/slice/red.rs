use leptos::{create_runtime, create_rw_signal};
use leptos_macro::slice;

#[derive(Default, PartialEq)]
pub struct OuterState {
    count: i32,
    inner: InnerState,
}

#[derive(Clone, PartialEq, Default)]
pub struct InnerState {
    inner_count: i32,
    inner_name: String,
}

fn main() {
    let _ = create_runtime();

    let outer_signal = create_rw_signal(OuterState::default());

    let (_, _) = slice!();

    let (_, _) = slice!(outer_signal);

    let (_, _) = slice!(outer_signal.);

    let (_, _) = slice!(outer_signal.inner.);
}
