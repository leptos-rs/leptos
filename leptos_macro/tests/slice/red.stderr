error: unexpected end of input, expected identifier
  --> tests/slice/red.rs:21:18
   |
21 |     let (_, _) = slice!();
   |                  ^^^^^^^^
   |
   = note: this error originates in the macro `slice` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected `.`
  --> tests/slice/red.rs:23:18
   |
23 |     let (_, _) = slice!(outer_signal);
   |                  ^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `slice` (in Nightly builds, run with -Z macro-backtrace for more info)

error: unexpected end of input, expected identifier or integer
  --> tests/slice/red.rs:25:18
   |
25 |     let (_, _) = slice!(outer_signal.);
   |                  ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `slice` (in Nightly builds, run with -Z macro-backtrace for more info)

error: unexpected end of input, expected identifier or integer
  --> tests/slice/red.rs:27:18
   |
27 |     let (_, _) = slice!(outer_signal.inner.);
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `slice` (in Nightly builds, run with -Z macro-backtrace for more info)
